pallets/capacity/src/lib.rs

```diff
@@ -518,19 +517,81 @@ impl<T: Config> Pallet<T> {
 		unstaker: &T::AccountId,
 		amount: BalanceOf<T>,
 	) -> Result<BalanceOf<T>, DispatchError> {
-		let mut staking_account =
-			Self::get_staking_account_for(unstaker).ok_or(Error::<T>::NotAStakingAccount)?;
-		ensure!(amount <= staking_account.active, Error::<T>::AmountToUnstakeExceedsAmountStaked);
+		StakingAccountLedger::<T>::try_mutate_exists(
+			unstaker,
+			|maybe_staking_account| -> Result<BalanceOf<T>, DispatchError> {
+				let mut staking_account =
+					maybe_staking_account.take().ok_or(Error::<T>::NotAStakingAccount)?;
+				ensure!(
+					amount <= staking_account.active,
+					Error::<T>::AmountToUnstakeExceedsAmountStaked
+				);
+
+				let actual_unstaked_amount = staking_account.withdraw(amount)?;
+				if staking_account.active.is_zero() {
+					*maybe_staking_account = None;
+				} else {
+					*maybe_staking_account = Some(staking_account);
+				}
+				Ok(actual_unstaked_amount)
+			},
+		)
+	}
 
+	fn add_unlock_chunk(
+		unstaker: &T::AccountId,
+		actual_unstaked_amount: BalanceOf<T>,
+	) -> Result<(), DispatchError> {
 		let current_epoch: T::EpochNumber = Self::get_current_epoch();
 		let thaw_at =
 			current_epoch.saturating_add(T::EpochNumber::from(T::UnstakingThawPeriod::get()));
+		let mut unlocks = Self::get_unstake_unlocking_for(unstaker).unwrap_or_default();
+		unlocks.add(actual_unstaked_amount, thaw_at)?;
+		UnstakeUnlocks::<T>::set(unstaker, Some(unlocks));
+		Ok(())
+	}
 
-		let unstake_result = staking_account.withdraw(amount, thaw_at)?;
-
-		Self::set_staking_account(&unstaker, &staking_account);
+	// Calculates a stakable amount from a proposed amount.
+	pub(crate) fn get_stakable_amount_for(
+		staker: &T::AccountId,
+		proposed_amount: BalanceOf<T>,
+	) -> BalanceOf<T> {
+		let account_balance = T::Currency::free_balance(&staker);
+		account_balance
+			.saturating_sub(T::MinimumTokenBalance::get())
+			.min(proposed_amount)
+	}
 
-		Ok(unstake_result)
+	pub(crate) fn do_withdraw_unstaked(
+		staker: &T::AccountId,
+	) -> Result<BalanceOf<T>, DispatchError> {
+		let current_epoch = Self::get_current_epoch();
+		let mut amount_withdrawn: BalanceOf<T> = Zero::zero();
+		let mut total_unlocking: BalanceOf<T> = Zero::zero();
+		UnstakeUnlocks::<T>::try_mutate_exists(
+			staker,
+			|maybe_unlocks| -> Result<(), DispatchError> {
+				let mut unlocks =
+					maybe_unlocks.take().ok_or(Error::<T>::NoUnstakedTokensAvailable)?;
+				amount_withdrawn = unlocks.reap_thawed(current_epoch);
+				ensure!(!amount_withdrawn.is_zero(), Error::<T>::NoUnstakedTokensAvailable);
+				if unlocks.unlocking.is_empty() {
+					*maybe_unlocks = None;
+				} else {
+					total_unlocking = unlocks.total();
+					*maybe_unlocks = Some(unlocks);
+				}
+				Ok(())
+			},
+		)?;
+		let staking_account = Self::get_staking_account_for(staker).unwrap_or_default();
+		let total_locked = staking_account.active.saturating_add(total_unlocking);
```

Comment:
`total_locked` will be zero if they've already completely unstaked (in which case staking_account.active is the default of 0), and all unstake unlocks have thawed (in which case `total_unlocking` never gets modified from 0).

Reply:
@aramikm note here `unwrap_or_default` to takes care of the case described above so staking_account.active would be 0.